### PR TITLE
Add basic markdown pre-process features

### PR DESCRIPTION
**Pull request by kjeller**
- Add a callback method for processing markdown metadata
- Add utility methods to output `.md`/`.kt` files to generated resources/sources directories

### CODE REVIEW

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KobwebxMarkdownPlugin.kt [hunk 1]
```diff
@@ -28,10 +29,14 @@ class KobwebxMarkdownPlugin : Plugin<Project> {
...
             .register<ConvertMarkdownTask>("kobwebxMarkdownConvert", kobwebBlock, markdownBlock)
 
+        val processTask = project.tasks
+            .register<ProcessMarkdownTask>("kobwebxMarkdownDataProcess", kobwebBlock, markdownBlock)
+
         project.buildTargets.withType<KotlinJsIrTarget>().configureEach {
             val jsTarget = JsTarget(this)
             project.kotlin.sourceSets.named(jsTarget.mainSourceSet) {
                 kotlin.srcDir(convertTask)
+                kotlin.srcDir(processTask)
```
Comment (DennisTsar) [inline tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KobwebxMarkdownPlugin.kt:39]:
```suggestion
                kotlin.srcDir(processTask.map { it.getGenDir() })
```
To ensure that only the kt files are treated as sources. 
I'd also probably rename `getGenDir()` to `genGenSrcDir` or `getGenKtDir` just to be more explicit.

**Review by bitspittle — COMMENTED** [review]
A handful of comments ahead, but a lot of them are documentation, because I assume you'd rather not write them :) However, feel free to treat all my doc suggestions as just that -- suggestions -- and choose the final wording you think is best.

This definitely feels like the correct shape of the feature to me, and I'm really happy to see it. Thanks for taking the time to create this PR -- honestly because of original issue's ambiguity I thought it might be one of the harder bugs to start with, but nah, you made short work of it.

Finally, if you haven't seen it before, "Nit:" means something trivial. It means I think it's worth mentioning but not the sort of thing I would lose sleep over if you disagreed with me. "Optional" is stuff I feel neutral about; thought I should say something but it doesn't really matter.

In general, I can sometimes leave a lot of comments, but I never want it to feel overwhelming, or like an attack. I want you to always feel comfortable pushing back, or asking for more clarification. We're all pursuing the best code together, here!

I'm looking forward to updating my blog site using these changes!

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KobwebxMarkdownPlugin.kt (hunk 1 above)
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KobwebxMarkdownPlugin.kt:39]: That's a good idea.
Let's go with `genGenSrcDir`.

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownBlock.kt [hunk 2]
```diff
@@ -79,6 +79,44 @@ abstract class MarkdownBlock(baseGenDir: Provider<String>) : KobwebBlock.FileGen
...
+     * potentially add an additional markdown file as a result of this call, although
+     * normally we expect users will just generate Kotlin files).
+     *
+     * @see MarkdownEntry
+     */
+    abstract val process: Property<ProcessScope.(List<MarkdownEntry>) -> Unit>
+
+    class ProcessScope {
+
+        data class ProcessNode(
```
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownBlock.kt:99]: I prefer OutputFile over ProcessNode, let's go with that.

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import org.commonmark.node.CustomBlock
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.getByType
+import java.io.File
+import javax.inject.Inject
+
+class MarkdownVisitor : AbstractVisitor() {
+    private val _frontMatter = mutableMapOf<String, List<String>>()
```
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt:26]: Yup! I copied this from my build.gradle.kts file which in turn was copied from yours. So I think this Visitor with its strict pattern comes from you originally 😃

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    }
+            )
+        }
+    }
+}
+
+abstract class ProcessMarkdownTask @Inject constructor(
+    private val kobwebBlock: KobwebBlock,
+    private val markdownBlock: MarkdownBlock,
+) : KobwebTask("Processes markdown files found in the project's resources path to metadata and provides methods for user to add custom generated files to the final project") {
```
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt:46]: Much more concise than my desc. Let's go with this one.

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            .toList()
+    }
+
+    @InputFiles
+    fun getMarkdownFiles(): List<File> {
+        return getMarkdownFilesWithRoots().map { it.file }
+    }
+
+    @OutputDirectory
+    fun getGenDir(): File = kobwebBlock.getGenJsSrcRoot<MarkdownBlock>(project).resolve(
```
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt:70]: Is it ok if I call it `genGenSrcDir` I am basing it on this comment https://github.com/varabyte/kobweb/pull/452#discussion_r1470389635

Then we have `genGenSrcDir` for Kt and `genGenResDir` for md

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            parser
+                .parse(it.readText())
+                .accept(visitor)
+            MarkdownEntry(
+                filePath = it.path,
+                frontMatter = visitor.frontMatter
+            )
+        }
+        val process = markdownBlock.process.orNull ?: return
+        val processScope = MarkdownBlock.ProcessScope()
```
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/tasks/ProcessMarkdownTask.kt:93]: Thank you! After you sent me your example usage it clicked for me. Then it was easier to understand all the previous comments👍

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownBlock.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownBlock.kt:99]: Sounds good! Also aligning parameter names in generate methods

File: tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownBlock.kt [hunk 7]
```diff
@@ -79,6 +79,44 @@ abstract class MarkdownBlock(baseGenDir: Provider<String>) : KobwebBlock.FileGen
...
+            val path: String,
```
[the remark this answers is not in the document]
Comment (kjeller) [reply tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownBlock.kt:100]: Aah nevermind I saw your other comment it is supposed to be `require(..)`

**Review by bitspittle — APPROVED** [review]
Perfect -- thanks!

I'll push out a snapshot shortly and test this later today.

**Pull request merged**